metatensor-torch/src/labels.cpp

```diff
@@ -339,55 +333,83 @@ TorchLabels LabelsHolder::permute(std::vector<int64_t> dimensions_indexes) const
-    auto names = this->names();
+    auto new_names = this->names();
 
-    auto it = std::find(std::begin(names), std::end(names), name);
-    if (it == std::end(names)) {
+    auto it = std::find(std::begin(new_names), std::end(new_names), name);
+    if (it == std::end(new_names)) {
         C10_THROW_ERROR(
             ValueError,
             "'" + name + "' not found in the dimensions of these Labels"
         );
     }
 
-    auto column_index = it - std::begin(names);
-    names.erase(it);
+    auto column_index = it - std::begin(new_names);
+    new_names.erase(it);
 
     auto values = this->values();
     auto first = values.index({torch::indexing::Slice(), torch::indexing::Slice(0, column_index)});
     auto second = values.index({torch::indexing::Slice(), torch::indexing::Slice(column_index + 1)});
     auto new_values = torch::hstack({first, second});
 
-    return torch::make_intrusive<LabelsHolder>(std::move(names), std::move(new_values));
+    return torch::make_intrusive<LabelsHolder>(std::move(new_names), std::move(new_values));
 }
 
 
 TorchLabels LabelsHolder::rename(std::string old_name, std::string new_name) const {
-    auto names = this->names();
+    auto new_names = this->names();
 
-    auto it = std::find(std::begin(names), std::end(names), old_name);
-    if (it == std::end(names)) {
+    auto it = std::find(std::begin(new_names), std::end(new_names), old_name);
+    if (it == std::end(new_names)) {
         C10_THROW_ERROR(
             ValueError,
             "'" + old_name + "' not found in the dimensions of these Labels"
         );
     }
-    auto column_index = it - std::begin(names);
+    auto column_index = it - std::begin(new_names);
 
-    names[column_index] = std::move(new_name);
-    return torch::make_intrusive<LabelsHolder>(std::move(names), std::move(this->values()));
+    new_names[column_index] = std::move(new_name);
+    return torch::make_intrusive<LabelsHolder>(std::move(new_names), this->values());
 }
 
 TorchLabels LabelsHolder::to(torch::IValue device) const {
     // transform torch::IValue into torch::Device
-    auto normalized_device = details::normalize_device(device);
-
-    // move the values
-    auto new_values = values_.to(normalized_device);
-
-    // copy the names
-    auto new_names = this->names();
+    auto normalized_device = torch::Device("cpu");
+    if (device.isString()) {
+        normalized_device = torch::Device(device.toStringRef());
+    } else if (device.isDevice()) {
+        normalized_device = device.toDevice();
+    } else {
+        C10_THROW_ERROR(TypeError,
+            "'device' must be a string or a torch.device, got '" + device.type()->str() + "' instead"
+        );
+    }
 
-    return torch::make_intrusive<LabelsHolder>(std::move(new_names), std::move(new_values));
+    if (normalized_device == values_.device()) {
+        // return the same object
+        return torch::make_intrusive<LabelsHolder>(*this);
```

Comment:
Do I understand this correctly? 

Before this commit a new LabelsHolder was returned in this function with the same memory location for the values, but without associating the lifetime of the old Rust labels values with the returned one.

Now with this commit uses instead the constructor
```
LabelsHolder::LabelsHolder(metatensor::Labels labels): labels_(std::move(labels)) {
```
(even though I don't understand how the types match for `LabelsHolder` and `metatensor::Labels` with `explicit` argument in the constructor)
which associates the lifetime of the existing Rust labels values with the returned LabelsHolder.

Reply:
yes, but this happens in the else branch. This branch just invoke the copy constructor of `LabelsHolder`.